lib/MonadicRewrite.thy

```diff
@@ -403,72 +522,88 @@ lemma monadic_rewrite_gets_the_walk:
   apply simp
   done
 
-lemma monadic_rewrite_alternatives:
-  "\<lbrakk> monadic_rewrite E F P a c; monadic_rewrite E F Q b d \<rbrakk>
-      \<Longrightarrow> monadic_rewrite E F (P and Q) (a \<sqinter> b) (c \<sqinter> d)"
-  by (auto simp: monadic_rewrite_def alternative_def)
+lemma monadic_rewrite_gets_l:
+  "(\<And>x. monadic_rewrite F E (P x) (g x) m)
+   \<Longrightarrow> monadic_rewrite F E (\<lambda>s. P (f s) s) (gets f >>= (\<lambda>x. g x)) m"
+  by (auto simp add: monadic_rewrite_def exec_gets)
 
-lemma monadic_rewrite_weaken2:
-  "monadic_rewrite F E P f g
-     \<Longrightarrow> monadic_rewrite F' E' ((\<lambda>_. (F \<longrightarrow> F') \<and> (E' \<longrightarrow> E)) and P) f g"
-  apply (rule monadic_rewrite_gen_asm)
-  apply (rule monadic_rewrite_weaken[where F=F and E=E])
-  apply auto
+lemma monadic_rewrite_gets_the_bind:
+  assumes mr: "(\<And>v. monadic_rewrite F E (Q v) (g v) m)"
+  shows   "monadic_rewrite F E (\<lambda>s. f s \<noteq> None \<and> Q (the (f s)) s) (gets_the f >>= (\<lambda>x. g x)) m"
```

Comment:
Nitpick: did you want the extra spaces after the `shows`?

Reply:
Thanks, a bunch of these are very old.